src/utils/Constants.hpp

```diff
@@ -64,4 +66,45 @@ std::string getFontPath()
 
 const std::string fontPath = getFontPath();
 
-const sf::Vector2u textPosition = {sideLength * (colSize + 2), sideLength * 2};
\ No newline at end of file
+const sf::Vector2u textPosition = {sideLength * (colSize + 2), sideLength * 2};
+
+std::string getPatternPath()
+{
+    const std::string relativeFontPath = "src/utils/pattern.txt";
+    return filePrefixPath + relativeFontPath;
+}
+
+const std::string patternPath = getPatternPath();
+
+std::vector<std::string> readPatternToString (std::string filename){
+    std::ifstream fin(filename);
+    if(!fin.is_open()){
+        std::cout << "File open failed, exit program" << std::endl;
+        exit(1);
+    }
+    std::string singleLine;
+    std::vector<std::string> allLines = {};
+    while(fin >> singleLine){
+        allLines.push_back(singleLine);
+    }
+    fin.close();
+    return allLines;
+}
+
+const std::vector<std::string> patternMatrixInString = readPatternToString(patternPath);
+
+std::vector<std::vector<State>> patternMatrixInStringToMatrix (const std::vector<std::string> &inputPattern){
+    std::vector<std::vector<State>> matrix(inputPattern.size(),std::vector<State>(inputPattern[0].size()));
+    for (int row = 0; row < (int)inputPattern.size(); row++){
```

Comment:
We should check if the matrix is empty first or directly call `inputPattern[0]` would fail the program if the pattern.txt is empty

Reply:
PS function naming convention: verb + noun

`convertPatternStringToStateMatrix` ?

Reply:
> We should check if the matrix is empty first or directly call `inputPattern[0]` would fail the program if the pattern.txt is empty

Thanks, you are right. I should check it here instead of when I'm calling the function and applying it to the button-click reaction